src/DFG.cpp

```diff
@@ -100,6 +133,95 @@ void DFG::tuneForPattern() {
+      for (DFGNode* predNode: *(dfgNode->getPredNodes())) {
+        // delete all edges between predNode and merged dfgNode
+        DFGEdge* firstEdge = NULL;
+        while (hasDFGEdge(predNode, dfgNode)){
+          if (firstEdge == NULL) {
+            firstEdge = getDFGEdge(predNode, dfgNode);
+          }
+          deleteDFGEdge(predNode, dfgNode);
+        }     
+        // add the first edge back
+        if (firstEdge){
+          m_DFGEdges.push_back(firstEdge);
+        }               
+      }
+    }
+  }
+}
+
+void DFG::tuneForDualIssue() {
+  // tuneForDualIssue reconstruct the edge for dual issue without tuneForPattern()
+  // the edges between the merged DFGNodes and its parents/children must be combined into one edge only.
+  // reconstruct connected DFG by modifying m_DFGEdge
+  std::cout<<"[MMJ] tuneForDualIssue is running " <<std::endl;
+  list<DFGNode*>* removeNodes = new list<DFGNode*>();
+  for (DFGNode* dfgNode: nodes) {
+    if (dfgNode->hasMerged() and !dfgNode->hasCombined()) {
+      if (dfgNode->isPatternRoot()) {
+        std::cout<<"[MMJ] dfgNode is " << dfgNode->getID() <<std::endl;
+        for (DFGNode* patternNode: *(dfgNode->getPatternNodes())) {
+          std::cout<<"[MMJ] patternNode is " << patternNode->getID() <<std::endl;
+          if (hasDFGEdge(dfgNode, patternNode))
+            m_DFGEdges.remove(getDFGEdge(dfgNode, patternNode));
+
+          for (DFGNode* predNode: *(patternNode->getPredNodes())) {
+            // we only keep one edge between dfgNode and predNode
+            if (predNode == dfgNode or
+                predNode->isOneOfThem(dfgNode->getPatternNodes())) {
+              deleteDFGEdge(predNode, patternNode);
+              continue;
+            }
+            DFGNode* newPredNode = NULL;
+            if (predNode->hasMerged() or predNode->hasCombined())
+              newPredNode = predNode->getPatternRoot();
+            else
+              newPredNode = predNode;
+            replaceDFGEdge(predNode, patternNode, newPredNode, dfgNode);              
+          }
+
+          for (DFGNode* succNode: *(patternNode->getSuccNodes())) {
+            std::cout<<"[MMJ] succNode is " << succNode->getID() <<std::endl;
+            if (succNode == dfgNode or
+                succNode->isOneOfThem(dfgNode->getPatternNodes())) {
+              deleteDFGEdge(patternNode, succNode);
+              continue;
+            }
+            DFGNode* newSuccNode = NULL;
+            if (succNode->hasMerged() or succNode->hasCombined())
+              newSuccNode = succNode->getPatternRoot();
+            else
+              newSuccNode = succNode;
+            replaceDFGEdge(patternNode, succNode, dfgNode, newSuccNode);
+          }
+        }
+      } else {
+        removeNodes->push_back(dfgNode);
+      }
+    }
+  }
+  for (DFGNode* dfgNode: *removeNodes) {
+    nodes.remove(dfgNode);
+  }
+}
+
+void DFG::ESCORT() {
+  dualIssueNew(2,100);
+  // If dualIssueNew fails
+  dualIssue(3,100);
```

Comment:
Not quite understand why we need two `dualIssue()` funcs. And not sure why there are two `tuneForDualIssue()` funcs. And what the difference between `dualissue` and `tuneFor`. Can you plz add comment here?